// livereactload/lib/browserify-plugin/main.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = LiveReactloadPlugin;
var through = require("through2");

var _require = require("./reloadServer");

var startServer = _require.startServer;
var makeHash = require("./makeHash");

var _require2 = require("fs");

var readFileSync = _require2.readFileSync;

var _require3 = require("path");

var resolve = _require3.resolve;
var dirname = _require3.dirname;
var basename = _require3.basename;

var _require4 = require("../common");

var values = _require4.values;

function LiveReactloadPlugin(b) {
  var opts = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
  var _opts$port = opts.port;
  var port = _opts$port === undefined ? 4474 : _opts$port;
  var _opts$host = opts.host;
  var host = _opts$host === undefined ? "localhost" : _opts$host;
  var _opts$client = opts.client;
  var client = _opts$client === undefined ? true : _opts$client;

  // server is alive as long as watchify is running
  var server = opts.server !== false ? startServer({ port: Number(port) }) : null;
  var requireOverride = readFileSync(resolve(__dirname, "../requireOverride.js")).toString();

  var clientOpts = {
    port: Number(port),
    host: host.toString()
  };

  b.on("reset", addHooks);
  addHooks();

  function addHooks() {
    // this cache object is preserved over single bundling
    // pipeline so when next bundling occurs, this cache
    // object is thrown away
    var modules = {};

    var bundleInitJs = "var require = " + requireOverride + ";\n\n       window.__livereactload$$ = {\n         require: require,\n         modules: modules,\n         exports: {},\n         reloaders: {},\n         initModules: initModules\n       };\n\n       initModules();\n\n       function initModules() {\n         var allExports = window.__livereactload$$.exports;\n         var modules    = window.__livereactload$$.modules;\n         // initialize Browserify compatibility\n         Object.keys(modules).forEach(function(id) {\n           modules[id][0] = (function(require, module, exports) {\n             if (!modules[id].__inited) {\n               modules[id].__inited = true\n               var __init = new Function(\"require\", \"module\", \"exports\", modules[id].source);\n               var _require = (function() { return require.apply(require, Array.prototype.slice.call(arguments).concat(id)); });\n               __init(_require, module, exports, arguments[3], arguments[4], arguments[5], arguments[6]);\n             }\n           })\n           modules[id][1] = modules[id].deps;\n         })\n       }";

    var originalEntry = "";
    var entryId = -1;

    // task of this hook is to override the default entry so that
    // the new entry
    b.pipeline.get("record").push(through.obj(function transform(row, enc, next) {
      var entry = row.entry;
      var file = row.file;

      if (entry) {
        originalEntry = file;
        next(null);
      } else {
        next(null, row);
      }
    }, function flush(next) {
      var origFilename = basename(originalEntry),
          origDirname = dirname(originalEntry);

      var newEntryPath = resolve(origDirname, "___livereactload_entry.js");
      var newEntrySource = [];

      if (client !== false) {
        var args = ["null", JSON.stringify(clientOpts)];
        if (client !== true) {
          var customClient = "require(" + JSON.stringify(client) + ", entryId$$)";
          args.push(customClient);
        }
        newEntrySource.push("require(\"livereactload/client\", entryId$$).call(" + args.join() + ");");
      }

      newEntrySource.push("require(" + JSON.stringify("./" + origFilename) + ", entryId$$);");

      this.push({
        entry: true,
        expose: false,
        file: newEntryPath,
        id: newEntryPath,
        source: newEntrySource.join("\n"),
        nomap: true,
        order: 0
      });
      next();
    }));

    b.pipeline.get("label").push(through.obj(function transform(row, enc, next) {
      var id = row.id;
      var index = row.index;
      var dedupeIndex = row.dedupeIndex;
      var file = row.file;
      var source = row.source;
      var deps = row.deps;
      var entry = row.entry;

      if (entry) {
        entryId = id;
      }
      modules[id] = {
        id: id,
        index: index,
        dedupeIndex: dedupeIndex,
        file: file,
        source: source,
        deps: deps,
        entry: entry,
        hash: makeHash(source)
      };
      next(null, row);
    }, function flush(next) {
      next();
    }));

    b.pipeline.get("wrap").push(through.obj(function transform(row, enc, next) {
      next(null);
    }, function flush(next) {
      var bundleSrc = "(function(modules, entryId$$) {\n             " + bundleInitJs + "\n             (function() {\n               " + modules[entryId].source + "\n             })();\n           })(" + JSON.stringify(modules, null, 2) + ", " + JSON.stringify(entryId) + ");";
      this.push(new Buffer(bundleSrc, "utf8"));
      if (server) {
        server.notifyReload({ modules: modules, entryId: entryId });
      }
      next();
    }));
  }
}

function isGlobalModule(moduleFilename) {
  // assuming that livereload package is in global mdule directory (node_modules)
  // and this file is in ./lib/babel-plugin folder
  return moduleFilename.indexOf(resolve(__dirname, '../../..')) !== -1;
}
module.exports = exports["default"];
